cpp/src/text/edit_distance.cu

```diff
@@ -108,16 +94,179 @@ struct calculate_compute_buffer_fn {
+  if (length1 == 0 || length2 == 0) {
+    if (lane_idx == 0) { *d_result = length1 == 0 ? length2 : length1; }
+    return;
+  }
+
+  // setup the 3 working vectors for this string
+  auto v0 = d_buffer;
+  auto v1 = v0 + length1 + 2;
+  auto v2 = v1 + length1 + 2;
+  if (lane_idx == 0) {
+    v0[0] = 0;  // first diagonal
+    v1[0] = 1;  // second diagonal
+    v1[1] = 1;
+  }
+
+  // utility for navigating the diagonal of the matrix of characters for the 2 strings
+  auto next_itr = [](Iterator sitr, cudf::size_type length, Iterator itr, cudf::size_type offset) {
+    if constexpr (cuda::std::is_pointer_v<Iterator>) {
+      itr = sitr - offset;  // ASCII iterator
+    } else {
+      auto const pos = sitr.position() - offset;  // minimizes character counting
+      itr += (pos >= 0) && (pos < length) ? (pos - itr.position()) : 0;
+    }
+    return itr;
+  };
+
+  // top-left of the matrix
+  // includes the diagonal one after the max(length1,length2) diagonal
+  for (auto idx = 0; idx < length1; ++idx, ++input1) {
+    auto const n = idx + 2;      // diagonal length
+    auto const a = n > length1;  // extra diagonal adjust
+
+    auto jdx = static_cast<cudf::size_type>(lane_idx);
+    auto it1 = input1;
+    auto it2 = input2;
+
+    auto tile_count = cudf::util::div_rounding_up_safe(n + 1, tile_size);
+    while (tile_count--) {
+      auto const offset = (jdx - 1);
+      // locate the 2 characters to compare along the diagonal
+      it1 = next_itr(input1, length1, it1, offset);
+      it2 = next_itr(input2, length2, it2, -offset);
+      if (jdx == 0) {
+        if (!a) { v2[0] = n; }
+      } else if (jdx < n) {
+        auto const sc = v0[jdx - 1] + (*it1 != *it2);
+        auto const dc = v1[jdx - 1] + 1;
+        auto const ic = v1[jdx] + 1;
+        v2[jdx - a]   = cuda::std::min(cuda::std::min(sc, dc), ic);
+      } else if (jdx == n) {
+        v2[n - a] = n;
+      }
+      tile.sync();
+      jdx += tile_size;
+    }
+    cuda::std::swap(v0, v1);
+    cuda::std::swap(v1, v2);
+  }
+
+  --input1;  // reset
+  ++input2;  // iterators
+
+  // bottom-right of the matrix
+  for (auto idx = 1; idx < length2; ++idx, ++input2) {
+    bool const fl = (length2 - idx) > length1;  // fill-last flag
+    auto const n  = (fl ? length1 : (length2 - idx)) + 1;
+
+    auto jdx = static_cast<cudf::size_type>(lane_idx);
+    auto it1 = input1;
+    auto it2 = input2;
+
+    auto tile_count = cudf::util::div_rounding_up_safe(n, tile_size);
+    while (tile_count--) {
+      auto const offset = (jdx - 1);
+      // locate the 2 characters to compare along the diagonal
+      it1 = next_itr(input1, length1, it1, offset);
+      it2 = next_itr(input2, length2, it2, -offset);
+      if (jdx > 0 && jdx < n) {
+        auto const sc = v0[jdx] + (*it1 != *it2);
+        auto const dc = v1[jdx - 1] + 1;
+        auto const ic = v1[jdx] + 1;
+        v2[jdx - 1]   = cuda::std::min(cuda::std::min(sc, dc), ic);
+      } else if (jdx == n && fl) {
```

Comment:
Very minor nit: It's a bit jarring that this conditional includes both conditions while in the top-left filling above the initial if has to do the `if (!a)` as a nested condition. I understand why (above if you don't do that then `jdx==0` will fall into the first else if condition instead), but maybe would help follow the code to mirror that pattern here.